packages/@glimmer/interfaces/lib/compile/wire-format.d.ts

```diff
@@ -219,6 +231,7 @@ export namespace Expressions {
     | GetPathFreeAsComponentOrHelperHeadOrThisFallback
     | GetPathFreeAsComponentOrHelperHead
     | GetPathFreeAsHelperHeadOrThisFallback
+    | GetPathFreeAsDeprecatedHelperHeadOrThisFallback
```

Comment:
Some of the `GetPathFree*` are not actually possible/used. Take `GetPathFreeAsHelperHeadOrThisFallback` for instance – if there is a path (i.e. `foo={{bar.baz}}` then `bar` must be a fallback for `this.bar` and cannot be a helper.

However, the current code, [specifically here](https://github.com/glimmerjs/glimmer-vm/blob/master/packages/@glimmer/compiler/lib/passes/2-encoding/expressions.ts#L93-L95) seems to require/assuming symmetry between `GetFree*` and `GetPathFree*`. Deleting the unused stuff from here seems to cause type issues.

In the end, I ended up abusing/hijacking the "path" portion of `GetPathFreeAsDeprecatedHelperHeadOrThisFallback` to save the argument name for the deprecation message (adding a `string` element to the tuple also breaks types for similar reasons).

Reply:
> if there is a path (i.e. foo={{bar.baz}} then bar must be a fallback for this.bar and cannot be a helper

Hmm, is that correct? I thought this should work (as of ember-source@3.25+), and is part of "everything is a value" conceptually. So both having a yielded block param for `bar` with a property that is `baz` that contains a `Ember.Helper.helper(() => {})` and having the same thing at `this.bar` **should** invoke that helper. 🤔

Reply:
Sure, but that would also not go through this path as bar is a local variable.

This is only used when bar is a “free variable” (what was called ambient scope in strict mode) _but_ we need to lookup bar as a resolved helper, and failing that, fallback to this.bar, which should never happen as far as I can tell.

Reply:
This is correct I believe, the path stuff for those `GetFree` opcodes was from before the implementation of strict mode, when we were still thinking of imported values as free variables. Now, imported values are in-scope values, which means that they never become any kind of `GetFree`.

Reply:
@pzuraq is correct